src/careamics/losses/lvae/losses.py

```diff
@@ -158,94 +189,134 @@ def _get_weighted_likelihood(
-        # if _restricted_kl:
-        #     pow = np.power(2,min(i + 1, _multiscale_count-1))
-        #     norm_factor /= pow * pow
-
+        # NOTE: we want to normalize the KL-loss w.r.t. the latent space dimensions,
+        # i.e., the number of entries in the latent space tensors (C, [Z], Y, X).
+        # We assume z has shape (B, C, [Z], Y, X), where `C = z_dims[i]`.
+        norm_factor = np.prod(topdown_data["z"][i].shape[1:])
         kl[:, i] = kl[:, i] / norm_factor
 
-    kl_loss = free_bits_kl(kl, 0.0).mean()
+    kl_loss = free_bits_kl(kl, 0.0).mean()  # shape: (1, )
+    # NOTE: free_bits disabled!
     return kl_loss
 
 
-def get_kl_divergence_loss(topdown_layer_data_dict, img_shape, kl_key="kl"):
-    # TODO explain all params!
-    """kl[i] for each i has length batch_size resulting kl shape: (bs, layers)."""
+def get_kl_divergence_loss_denoisplit(
+    topdown_data: dict[str, torch.Tensor],
+    img_shape: tuple[int],
+    kl_key: str = "kl",
+) -> torch.Tensor:
+    """Compute the KL divergence loss for denoiSplit.
+
+    Parameters
+    ----------
+    topdown_data : dict[str, torch.Tensor]
+        A dictionary containing information computed for each layer during the top-down
+        pass. The dictionary must include the following keys:
+        - "kl": The KL-loss values for each layer. Shape of each tensor is (B,).
+        - "z": The sampled latents for each layer. Shape of each tensor is
+        (B, layers, `z_dims[i]`, H, W).
+    img_shape : tuple[int]
+        The shape of the input image to the LVAE model. Shape is ([Z], Y, X).
+    kl_key : str
+        The key for the KL-loss values in the top-down layer data dictionary.
+        To choose among ["kl", "kl_restricted", "kl_spatial", "kl_channelwise"]
+        Default is "kl"
+
+    kl[i] for each i has length batch_size resulting kl shape: (bs, layers).
+    """
     kl = torch.cat(
-        [kl_layer.unsqueeze(1) for kl_layer in topdown_layer_data_dict[kl_key]],
+        [kl_layer.unsqueeze(1) for kl_layer in topdown_data[kl_key]],
         dim=1,
     )
 
-    # As compared to uSplit kl divergence,
-    # more by a factor of 4 just because we do sum and not mean.
-    kl_loss = free_bits_kl(kl, 1).sum()  # TODO hardcoded free_bits 1
-    # NOTE: at each hierarchy, it is more by a factor of 128/i**2).
+    kl_loss = free_bits_kl(kl, 1.0).sum()
+    # NOTE: as compared to uSplit kl divergence, this KL loss is larger by a factor of
+    # `n_layers` since we sum KL contributions from different layers instead of taking
+    # the mean.
+
+    # NOTE: at each hierarchy, the KL loss is larger by a factor of (128/i**2).
     # 128/(2*2) = 32 (bottommost layer)
     # 128/(4*4) = 8
     # 128/(8*8) = 2
     # 128/(16*16) = 0.5 (topmost layer)
 
-    # Normalize the KL-loss w.r.t. the  latent space
+    # Normalize the KL-loss w.r.t. the input image spatial dimensions (e.g., 64x64)
     kl_loss = kl_loss / np.prod(img_shape)
```

Comment:
At this location, we additionally need to divide by number of z dimensions for the 3D case. but nothing to be done now.

Reply:
Do you mean that we need to divide by (Z*Y*X)? If this is the case I think it is already suitable for this, as in the `denoisplit_loss` `img_shape=targets[2:]` (i.e., all the spatial dimensions, supposing that the shape of `target` is `(B, C, [Z], Y, X)`.

Reply:
Makes sense.